crates/sui-core/src/state_accumulator.rs

```diff
@@ -458,71 +602,166 @@ impl StateAccumulator {
+        let epoch = epoch_store.epoch();
+        let running_root = epoch_store
+            .get_highest_running_root_accumulator()
+            .expect("Failed to get running root accumulator from disk");
+        let running_root_data = match (epoch, running_root) {
+            // running from genesis - start from scratch
+            (0, None) => None,
+            // reconfig || snapshot restore - start from previous root state hash
+            (epoch, None) => {
+                let (highest_acc, running_root) = store
+                    .get_root_state_accumulator_for_epoch(epoch - 1)
+                    .expect("Failed to get root state accumulator for previous epoch")
+                    .expect("Expected root state accumulator for previous epoch to exist");
+                Some((highest_acc, running_root.clone()))
+            }
+            // Mid-epoch restart - load from disk
+            (_, Some((highest_acc, running_root))) => Some((highest_acc, running_root)),
+        };
+        Self {
+            store,
+            running_root_accumulator: Mutex::new(running_root_data),
+        }
     }
 
-    pub fn accumulate_live_object(acc: &mut Accumulator, live_object: &LiveObject) {
-        match live_object {
-            LiveObject::Normal(object) => {
-                acc.insert(object.compute_object_reference().2);
+    pub async fn accumulate_running_root(
+        &self,
+        epoch_store: &AuthorityPerEpochStore,
+        checkpoint_seq_num: CheckpointSequenceNumber,
+        checkpoint_acc: Option<Accumulator>,
+    ) -> SuiResult {
+        let _scope = monitored_scope("AccumulateRunningRootV2");
+        // first check if we need to wait for other checkpoints to be accumulated. This can happen for the end
+        // of epoch checkpoint. Create a new scope to avoid deadlocking while we wait
+        let await_checkpoints = {
+            match self.running_root_accumulator.lock().as_ref() {
+                Some((highest_accumulated, _)) if *highest_accumulated < checkpoint_seq_num - 1 => {
+                    // TODO: take the checkpoint and check for EndOfEpochData to confirm we're at
+                    // the end of the epoch
+                    true
+                }
+                Some(_) => false,
+                None => false,
             }
-            LiveObject::Wrapped(key) => {
-                acc.insert(
-                    bcs::to_bytes(&WrappedObject::new(key.0, key.1))
-                        .expect("Failed to serialize WrappedObject"),
+        };
+        if await_checkpoints {
+            debug!(
+                "Awaiting accumulation up to checkpoint {} for epoch {} before adding to running root",
+                checkpoint_seq_num - 1,
+                epoch_store.epoch()
+            );
+            epoch_store
+                .notify_read_running_root(checkpoint_seq_num - 1)
+                .await
+                .expect("Failed to notify read running root");
+        }
```

Comment:
Is the code above just an optimization to avoid calling notify_read_running_root sometimes? It seems like it would be much simpler to simply always call it (unless checkpoint_seq_num is 0). I don't think this code will be even close to being a performance problem.

Reply:
i'm also wondering if we even need the in-memory state now that we store the running accumulator with a seq number. It seems like this method could be as simple as
```
let mut running_root = if checkpoint_seq_num == 0 { 
  Accumulator::default()
} else {
  epoch_store.notify_read_running_root(checkpoint_seq_num - 1)
};
running_root.union(checkpoint_acc);
epoch_store.insert_running_root_accumulator(&checkpoint_seq_num, running_root)?;
```

Reply:
Good point. The per checkpoint running root was a last minute change to allow for introducing a notify read pattern which I needed for checkpoint builder to wait at end of epoch. Removed the in-memory running root now and simplified (there's one edge case I needed to add to your above, to handle beginning of new epoch, at which point previous running root would not exist due to new epoch store)